proposals/p2200.md

```diff
@@ -0,0 +1,1047 @@
+if it is valid.
+
+#### Qualified member name lookup
+
+Adding qualifier to the member name, as in `T.(U.X)`, can make the lookup less
+dependent on the template parameter:
+
+-   If `T` is dependent, and `U` is an interface, the value of the expression is
+    dependent, but the type of the expression is not.
+    -   If `T` is known to implement `U`, then the lookup is not dependent. For
+        example, there could be a requirement on `T` or a sufficiently general
+        implementation of `U` that includes all possible values of `T`.
+    -   Otherwise, the lookup is template validity dependent.
+
+```carbon
+interface Hashable {
+  let HashType:! Type;
+  fn Hash[me: Self]() -> HashType;
+}
+
+external impl forall [T:! Serializable] T as Hashable;
+
+fn F[template T:! Serializable](x: T) {
+  // `T` is required to implement `Serializable` so this
+  // is not dependent.
+  x.(Serializable.Serialize)();
+
+  // Any `T` implementing `Serializable` also implements
+  // `Hashable`, since there is a blanket implementation,
+  // so this is not dependent. Note: there may be a
+  // specialization of this impl, so we can't rely on
+  // knowing how `T` implements `Hashable`.
+  x.(Hashable.Hash)();
+
+  // Unclear whether `T` implements `Printable`, but if
+  // does, clear what this means, so this is template
+  // validity dependent
+  x.(Printable.Print)();
+
+  match (x.(Hashable.Hash)()) {
+    // Uses the value of the associated type
+    // `Hashable.HashType` that is template dependent.
+    case _: u64 => { ... }
+    default => { ... }
+  }
+}
+```
+
+-   If `U.X` is dependent, then the entire expression is dependent.
+
+#### Impl lookup
+
+If the validity of an expression requires that an impl exist for a type, and
+that can't be determined until the value of a template parameter is known, then
+the expression is template validity dependent. For example, in the example from
+the previous section, `x.(Printable.Print)()` is valid if `T` implements
+`Printable`. This can also occur without a qualified lookup, for example:
+
+```carbon
+fn F[T:! Printable](x: T);
+fn G[template T:! Type](x: T) {
+  // Valid if `T` implements `Printable`, so this
+  // expression is template validity dependent.
+  F(x);
+}
+```
+
+The values of members of an impl for a template-dependent type are template
+dependent, unless they can be resolved to a not template-dependent expression
+using checked-generic impl resolution.
+
+```
+final external impl [T:! Type] T* as D
+    where .Result = T and .Index = i32;
+fn F[T:! Type](p: T*) {
+  // `(T*).(D.Index)` uses the final impl of `D` for `T*`,
+  // and so equals `i32`, which is not dependent.
+  // `(T*).(D.Result)` is recognized as equal to `T`, which
+  // is template dependent.
+}
+```
+
+#### Use of dependent value is dependent
+
+To match the expectations of C++ templates, uses of dependent values are also
+template dependent. For example:
```

Comment:
The example below is a bit unusual: we propagate the dependence of the case condition down (and sideways) into everything in the body. The normal case would be propagating dependence upwards, from subexpressions to enclosing expressions. Maybe we should give an example of that case first, and then discuss the `match` case as a special case?